### ISSUE TITLE
GitHub Action to install `cargo-run-bin`

### ISSUE THREAD

**Issue by lgarron** [dustinblackman/cargo-run-bin#21]
I really like the collective power of `cargo-run-bin` and https://github.com/cargo-bins/cargo-binstall , which can work well together. However, I'm having issues using them efficiently in GitHub Actions, for a few reasons:

- `cargo-run-bin` doesn't seem to have binaries that can be used in GitHub Actions. https://github.com/cargo-bins/cargo-binstall/issues/1551
- I think https://github.com/Swatinem/rust-cache messes with caching and can result in non-deterministic failures ([here](https://github.com/cubing/cubing.rs/actions/runs/7519349374/job/20467775758), for example, although the logs aren't public). Somehow it thinks the binary is installed when it's not?
- `cargo install cargo-run-bin` takes significant time, potentially dominating the runtime of a CI job.

This is probably "my fault" for not understanding exactly how to connect all the tooling but I think this could be a lot more straightforward. It would be really nice to have a GitHub action that installs a pre-compiled `cargo-run-bin` in a couple of seconds, similar to [`cargo-binstall`](https://github.com/cargo-bins/cargo-binstall). This would provide a great foundation for writing fast and repeatable CI jobs for Rust.

**Comment by lgarron** [comment]
For example, here are two CI runs of the exact same commit:

<img src="https://github.com/dustinblackman/cargo-run-bin/assets/248078/8b3351b4-09c5-4e8e-b935-988a31b47e7f" width="400"><img src="https://github.com/dustinblackman/cargo-run-bin/assets/248078/282c9308-7ba5-4c97-bd93-b7b6c537aab3" width="400">

I'm sure there are some details I'm missing, but one CI run takes only 29s while the other one spends 26s just on installing `cargo-run-bin`. Even if the rest of the CI run can take longer, it's especially valuable for the tooling to install super quickly, so that any failure is caught as quickly as possible. This can make the difference between seeing a CI failure before submitting a PR to one's project, vs. after.

To be clear, I'm not saying `cargo-run-bin` is doing anything wrong, I'd just really love to see everyone benefit from super fast CI jobs in GitHub Actions using `cargo-run-bin` in a standard/officially supported way.

**Comment by NobodyXu** [comment]
There's also quickinstall, which will build cargo-run-bin if it is used widely.

Or I can add it to the list of popular softwares to build, if you don't feel like adding binaries to release.

**Comment by dustinblackman** [comment]
Hey there!

So I'm down to build binaries for cargo-run-bin, I just need to put in the effort. I take producing third party licenses pretty seriously, and just need to port stuff from other repos first to get it setup.

`Swatinem/rust-cache` should be able to cache global dependency builds in `~/.cargo`, you may just have to selectively state you want it cached. I thought I had it setup on some other projects, but I've missed it and never noticed the run times.

Either way, binary releases sound fun. I'll try to make time.

**Comment by NobodyXu** [comment]
Fun fact:

@lgarron cargo-quickinstall already builds cargo-run-bin v1.7.2 https://github.com/cargo-bins/cargo-quickinstall/releases/tag/cargo-run-bin-1.7.2

So if you are using cargo-binstall or taiki-e/install-action, it should work out-of-the-box.

**Comment by lgarron** [comment]
> Fun fact:
> 
> @lgarron cargo-quickinstall already builds cargo-run-bin v1.7.2 https://github.com/cargo-bins/cargo-quickinstall/releases/tag/cargo-run-bin-1.7.2
> 
> So if you are using cargo-binstall or taiki-e/install-action, it should work out-of-the-box.

Do you have an example of an action that reliably installs `cargo-binstall` and `cargo-run-bin` quickly and consistently? I have yet to find a combination that works without causing some sort of caching issues.

**Comment by NobodyXu** [comment]
> Do you have an example of an workflow that reliably installs cargo-binstall and cargo-run-bin quickly and consistently when Swatinem/rust-cache is also in use? I have yet to find a combination that works without causing

Can you try installing cargo-run-bin first using taiki-e/install-action, then using the rust-cache?

**Comment by dustinblackman** [comment]
Nearly a year since this was opened, I'm sorry that I haven't gotten around it.

Instead of providing a Github Action, I'd recommend checking out the new wrapper instructions from @zerotacg instead. I feel like it'll provide the smoothest on-boarding experience for team members once it's setup. 

https://github.com/dustinblackman/cargo-run-bin/tree/master?tab=readme-ov-file#using-wrapper

**Comment by lgarron** [comment]
I gave that a go, and it borked immediately. :-/

```
╭─── /Users/lgarron/Code/git/github.com/cubing/twsearch/tools/cargo-bin ───────────────────────────────────────────────────────────────────────────────────
├─  cargo-run-bin-hack  | ✚3 
├─ lgarron@Germain
│ cargo bin --version
┴
warning: user-defined alias `bin` is shadowing an external subcommand found at: `/Users/lgarron/.cache/cargo/bin/cargo-bin`
This was previously accepted but is being phased out; it will become a hard error in a future release.
For more information, see issue #10049 <https://github.com/rust-lang/cargo/issues/10049>.
error: failed to parse manifest at `/Users/lgarron/Code/git/github.com/cubing/twsearch/tools/cargo-bin/Cargo.toml`

Caused by:
  error inheriting `edition` from workspace root manifest's `workspace.package.edition`

Caused by:
  `workspace.package.edition` was not defined
┬
├─ ❌ [101] command status
╰─── ⏱️ 0.042s ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Comment by lgarron** [comment]
Also, it seems that:

1. Giving the alias a non-conflicting name[^1]

**Comment by lgarron** [comment]
Also, it seems that:

1. Giving the alias a non-conflicting name[^1] shows `error inheriting `edition` from workspace root manifest's \`workspace.package.edition\`` even though that field is set.
2. The instructions seem to have a typo, invoking the package `cargo-bin` rather than `cargo-run-bin`.

This makes me feel the setup would be rather fragile even if it worked.

[^1]: `runbin` in my case, because I can't figure out how to set `run-bin`?

**Comment by lgarron** [comment]
Okay, I think I figured out some stuff: https://github.com/cubing/twsearch/compare/cargo-run-bin-hack

But no matter what I try, I can't figure out how to get the invoked tool to actually accept any passed command-line arguments (including the name of the package to run).

**Comment by lgarron** [comment]
Okay, I think the culprit is this: https://github.com/dustinblackman/cargo-run-bin/blob/9fc27205e49ff80d565a03fe437f66a2b83acdd9/src/cli.rs#L133

This means that global `cargo-bin` is incompatible with local `cargo-bin`, which is sort of good but also precludes using `cargo bin` for projects that are not configured with the local `cargo-bin` hack. It also forces a conflict with [`cargo-bin`](https://crates.io/crates/cargo-bin), which I think makes for bad ergonomics and semantics.

Would you be open to removing the binary name sniffing constraint?

**Comment by lgarron** [comment]
also btw, the code example fails Clippy linting:

![Image](https://github.com/user-attachments/assets/02702566-4628-4241-a4ec-cd6cfa636f58)

**Comment by zerotacg** [comment]
It looks like you pretty much have figured everything out already.
As you highlighted the alias collides with a global installation. Having the wrapper removes the need for a global installation.
So you can just replace the global one with the wrapper.
https://github.com/cubing/twsearch/compare/main...zerotacg:twsearch:main

Alternatively as you suggested you can use a different alias to be able to use both.
No matter which you choose you might still want to cache the folder `.bin` folder

**Comment by dustinblackman** [comment]
As you've found a working solution, I'll close this issue. Thanks for the effort! If an alternative is ever required, I've personally moved my projects (both public and private) to use the wrapper method. When onboarding new people on to a project, I've found it easier as it's one less step. https://github.com/dustinblackman/cargo-run-bin?tab=readme-ov-file#using-wrapper

**Comment by lgarron** [comment]
Yeah, the wrapper approach has a bunch of issues, but I think it's probably the least funky one until `cargo` implements something proper (which I think is years overdue, but here we are).

It would be nice to still be able to run crates as commands without having to do the dance to add them to a project locally (in particular, to try them out), but that's out of scope for this issue.
